hibernate-core/src/test/java/org/hibernate/test/proxy/HibernateUnproxyTest.java

```diff
@@ -25,6 +27,41 @@
         return new Class<?>[]{Parent.class, Child.class};
     }
 
+    @Test
+    public void testInitializedProxySubChildCanBeUnproxied() {
+        Parent u = new Parent();
+        SubChild c = new SubChild();
+        u.setChild(c);
+        doInJPA(this::entityManagerFactory, (entityManager -> {
+            entityManager.persist(u);
+        }));
+        doInJPA(this::entityManagerFactory, (entityManager -> {
+            Parent parent = entityManager.find(Parent.class, u.getId());
+            SubChild unproxiedChild = initializeAndUnproxy(getUser(parent));
```

Comment:
This test apparently also fails for Java 8 so this is not a true reproducer for HHH-12480. The way that the generic type for `getUser` is defined, it tries to cast `p.getChild()`, the proxy, to a `SubChild` already. This can't be done as the entity is not  initialised yet. At the time this proxy is made, only the id of the child is known, as well as its super type (i.e. the target type of the child association, which is in this case supposedly `Child`). To know the exact subtype you basically have to initialise the entire sub-child, eliminating the use for lazy loading.

You should cast to the subtype only after extracting the initialised proxy implementation in initializeAndUnproxy.

Reply:
@JWGmeligMeyling thank you for your response. When I run `./gradlew buiid` locally this is what I get:
```
> Task :hibernate-core:test

org.hibernate.test.proxy.HibernateUnproxyTest > testInitializedProxySubChildCanBeUnproxied FAILED
    java.lang.ClassCastException at HibernateUnproxyTest.java:40

org.hibernate.test.fileimport.CommandExtractorServiceTest > testImportFile FAILED
    org.junit.ComparisonFailure

org.hibernate.test.fileimport.MultiLineImportFileTest > testImportFile FAILED
    org.junit.ComparisonFailure at MultiLineImportFileTest.java:66

6422 tests completed, 3 failed, 350 skipped

> Task :hibernate-core:test FAILED

FAILURE: Build failed with an exception.

```

And I don't understand why pretty much the same code works fine with java8 and fails with java9 when I use the hibernate test-case templates:
https://github.com/ptahchiev/hibernate-test-case-templates/tree/HHH-12480

Reply:
It doesn't seem to work in Java 8 either and it doesn't fail where you expect it to fail.

Try with:

```java
    public Child getUser(Parent p) {
        return p.getChild();  // Get rid of cast to SubChild, as the proxy will be of type Child
    }

    public <T> T initializeAndUnproxy(Object entity) {
        if (entity == null) {
            throw new NullPointerException("Entity passed for initialization is null!");
        }

        if (!Hibernate.isInitialized(entity)) {
            Hibernate.initialize(entity);
            if (entity instanceof HibernateProxy) {
                entity = (T) ((HibernateProxy) entity).getHibernateLazyInitializer().getImplementation();
            }
        }
        return entity;
    }
```